lib/rtcpeerconnection/safari.js

```diff
@@ -236,47 +231,58 @@ SafariRTCPeerConnection.prototype.close = function close() {
   });
 };
 
-SafariRTCPeerConnection.prototype.addStream = function addStream(stream) {
-  if (this._localStreams.has(stream)) {
-    return;
+// NOTE(mmalavalli): webrtc-adapter has a bug where the "addTrack" shim
+// does not return an RTCRtpSender and returns undefined instead. An issue
+// [https://github.com/webrtc/adapter/issues/714] has been filed. For now,
+// we manually get the RTCRtpSender associated with the added track and
+// return it.
+//
+// UPDATE(mmalavalli): Because we are not delegating to the native
+// RTCPeerConnection#removeTrack(), we have to manually maintain a list of added
+// tracks. So we disable the delegation to the native RTCPeerConnection#addTrack()
+// for now.
+SafariRTCPeerConnection.prototype.addTrack = function addTrack() {
+  var args = [].slice.call(arguments);
+  var track = args[0];
+  var sender = this._senders.get(track.id);
+  if (sender && sender.track === track) {
+    throw new Error('Cannot add MediaStreamTrack [' + track.id + ', '
+      + track.kind + ']: RTCPeerConnection already has it');
   }
-  var tracks = new Set(stream.getTracks());
-  this._peerConnection.getSenders().forEach(function(sender) {
-    if (tracks.has(sender.track)) {
-      tracks.delete(sender.track);
-    }
-  });
-  tracks.forEach(function(track) {
-    try {
-      this.addTrack(track, stream);
-    } catch (error) {
-      // Do nothing.
-    }
-  }, this);
-  this._localStreams.set(stream, new Set(tracks));
-};
+  sender = getActiveSenders(this._peerConnection).get(track.id)
+    || this._peerConnection.addTrack.apply(this._peerConnection, args);
 
-SafariRTCPeerConnection.prototype.removeStream = function removeStream(stream) {
-  // NOTE(mroberts): We can't really remove tracks right now, at least if we
-  // ever want to add them back...
-  //
-  //     https://bugs.webkit.org/show_bug.cgi?id=174327
-  //
-  // var tracks = this._localStreams.get(stream) || new Set();
-  // this.getSenders().forEach(function(sender) {
-  //   if (tracks.has(sender.track)) {
-  //     this.removeTrack(sender);
-  //   }
-  // }, this);
-  this._localStreams.delete(stream);
+  sender = sender || getActiveSenders(this._peerConnection).get(track.id);
+  this._senders.set(track.id, sender);
```

Comment:
What's going on here?

Reply:
@markandrus Because of this issue in adapter.js [https://github.com/webrtc/adapter/issues/714] , their shimmed `addTrack` does not return the sender that was added, so we make sure we return the sender by manually getting it ourselves.